test/test-lex.cpp

```diff
@@ -339,6 +339,25 @@ TEST_F(test_lex, lex_binary_numbers) {
   this->check_tokens(u8"0b01n"_sv, {token_type::number});
 }
 
+TEST_F(test_lex, fail_lex_integer_loses_precision) {
+  this->check_tokens_with_errors(
+      u8"9007199254740993"_sv, {token_type::number},
+      [](padded_string_view input, const auto& errors) {
+        EXPECT_THAT(errors,
+                    ElementsAre(DIAG_TYPE_OFFSETS(
+                        input, diag_number_literal_will_lose_precision,  //
+                        characters, 0, u8"9007199254740993")));
```

Comment:
We should also test `rounded_val` (which means we need to use `DIAG_TYPE_2_FIELDS` instead of `DIAG_TYPE_2_OFFSETS`):
```suggestion
                    ElementsAre(DIAG_TYPE_2_FIELDS(
                        diag_number_literal_will_lose_precision,  //
                        characters, offsets_matcher(input, 0, u8"9007199254740993"), //
                        rounded_val, u8"9007199254740992"sv)));
```